tools/release/ltsgraph/springlayout.cpp

```diff
@@ -146,49 +177,68 @@ void SpringLayout::apply()
       }
       m_sforces[n] = (this->*m_forceCalculation)(m_graph.node(n).pos(), m_graph.stateLabel(n).pos(), 0.0);
     }
+    timer.endTiming("node force computation");
+    timer.startTiming("edge force computation");
 
     for (std::size_t i = 0; i < edgeCount; ++i)
     {
       std::size_t n = sel ? m_graph.explorationEdge(i) : i;
 
-      Edge e = m_graph.edge(n);
       QVector3D f;
-      // Variables for repulsion calculations
+      Edge e = m_graph.edge(n);
+      Node& handle = m_graph.handle(n);
+      NodeNode& node_from = m_graph.node(e.from());
+      NodeNode& node_to = m_graph.node(e.to());
 
+      // Variables for repulsion calculations
       m_hforces[n] = QVector3D(0, 0, 0);
       m_lforces[n] = QVector3D(0, 0, 0);
 
       if (e.is_selfloop())
       {
-        m_hforces[n] += repulsionForce(m_graph.handle(n).pos(), m_graph.node(e.from()).pos(), m_repulsion, m_natLength);
+        m_hforces[n] += repulsionForce(handle.pos(), node_from.pos(), m_repulsion, m_natLength);
       }
 
-      f = (this->*m_forceCalculation)(m_graph.node(e.to()).pos(), m_graph.node(e.from()).pos(), m_natLength);
+      f = (this->*m_forceCalculation)(node_to.pos(), node_from.pos(), m_natLength);
       m_nforces[e.from()] += f;
       m_nforces[e.to()] -= f;
 
-      f = (this->*m_forceCalculation)((m_graph.node(e.to()).pos() + m_graph.node(e.from()).pos()) / 2.0, m_graph.handle(n).pos(), 0.0);
+      f = (this->*m_forceCalculation)((node_to.pos() + node_from.pos()) / 2.0, handle.pos(), 0.0);
       m_hforces[n] += f;
 
-      f = (this->*m_forceCalculation)(m_graph.handle(n).pos(), m_graph.transitionLabel(n).pos(), 0.0);
+      f = (this->*m_forceCalculation)(handle.pos(), m_graph.transitionLabel(n).pos(), 0.0);
       m_lforces[n] += f;
 
-      for (std::size_t j = 0; j < i; ++j)
+      // for both nodes on this edge
+      for (std::size_t node : {e.from(), e.to()})
```

Comment:
What if `e.from()` and `e.to()` are the same node? I don't think you want to consider the same node twice. Otherwise this is a brilliant way to gain performance. :smile: